consensus/merge/merge.go

```diff
@@ -159,20 +160,49 @@ func (s *Merge) Finalize(config *chain.Config, header *types.Header, state *stat
 		}
 	}
 
-	return txs, r, nil
+	var rs types.Requests
+	if config.IsPrague(header.Time) {
```

Comment:
Currently `Finalize` looks messy. It takes some requests, verifies some fraction of them (and some verification is also done by `core.FinalizeBlockExecution`), adds some new requests and returns them 😖 And in most places the returned requests are ignored.
 
I suggest a cleaner separation of concerns:
- `FinalizeAndAssemble` is used for producing new blocks. It doesn't take `requests` as an input, but instead calculates and returns them (both deposits & withdrawals).
- `Finalize` is used for already mined blocks, so it can take `requests` as an input and only verify that they match  `RequestsRoot` (and thus the same bit of logic can be removed from `core.FinalizeBlockExecution`). It does not return requests.

Reply:
That would make for duplicated code because `misc.DequeueWithdrawalRequests7002()` needs to be `syscall`ed and that changes state. That should happen only once.
It's true that only `FinalizeAndAssemble` consumes the returned `requests` but, this is not new, most places ignore most returned results of `Finalize` so, it's fine i.m.o

Reply:
I wouldn't want to put consensus-linked code (e.g. any EIP-specific changes for, say Ethereum) into `core.FinalizeBlockExecution`

Reply:
Why does it have to be duplicated? I'm suggesting to move the `DequeueWithdrawalRequests7002` bit from `Finalize` to `FinalizeAndAssemble` and remove the `if !reflect.DeepEqual(rss, requests)`  check from `core.FinalizeBlockExecution`. So `Finalize` doesn't return any requests anymore and becomes something like
```
func (s *Merge) Finalize(..., requests types.Requests, ...
) (types.Transactions, types.Receipts, error) {
...
if config.IsPrague(header.Time) {
  rh := types.DeriveSha(requests)
  if header.RequestsRoot == nil || *header.RequestsRoot != rh {
    return nil, nil, fmt.Errorf("error: invalid requests root hash in header, expected: %v, got :%v", header.RequestsRoot, rh)
  }
}
return txs, receipts, nil
}
```
while `FinalizeAndAssemble` becomes something like
```
func (s *Merge) FinalizeAndAssemble(...
) (types.Transactions, types.Receipts, types.Requests, error) {
	if !misc.IsPoSHeader(header) {
		return s.eth1Engine.FinalizeAndAssemble(...)
	}
	outTxs, outReceipts, err := s.Finalize(...)
	if err != nil {
		return nil, nil, nil, err
	}
        var rs types.Requests
	if config.IsPrague(header.Time) {
		rs = ParseDepositL.ogs(receipts)
                rs = append(rs, misc.DequeueWithdrawalRequests7002(syscall)...)
	}
	return types.NewBlock(header, outTxs, uncles, outReceipts, withdrawals, rs), outTxs, outReceipts, rs, nil
}
```

Reply:
Maybe we don't even need the `RequestsRoot` check in `Finalize` because it's already done by `(b *Block) HashCheck()` or by `turbo/stages/bodydownload/body_algos.go` (see PR #10436).

Reply:
Actually, I forgot that EIP-6110 requires us to verify that deposit requests in a block match expected deposit requests, so we need to keep that logic in `Finalize`. But then why doesn't EIP-7002 require the same? Looks like an omission from the spec.